[#55968] Implement project removal with warning dialog before removing the project

judithroth opened the pull request:
https://community.openproject.org/work_packages/55968

judithroth committed:
[#55968] Implement project removal with warning dialog before removing the project

https://community.openproject.org/work_packages/55968

judithroth committed:
Merge branch 'dev' into implementation/55968-implement-project-removal-with-warning-dialog-before-removing-the-project

judithroth committed:
Try to satisfy rubocop on CI (remove unnecessary rubocop:disable)

judithroth committed:
Use correct way to require the module's spec helper

akabiru commented on modules/storages/app/controllers/storages/admin/storages/project_storages_controller.rb:
```diff
@@ -73,10 +72,24 @@ def create
 
-  def destroy; end
+  def destroy_confirmation_dialog
+    respond_with_dialog Storages::ProjectStorages::DestroyConfirmationDialogComponent.new(
```
🟢 Q: Do we still need an async dialog for this use case? Given the use case does not require streamable forms. A "vanilla" dialog would perhaps be simpler and more fitting

judithroth replied:
In general I agree with you and think from the contents of the dialog it does not need to be async. However, I didn't want to change the way the `action_menu` of a table row works and this currently only works with links. So as long as there is no pressing reason why the dialog should be embedded in the page, I'd avoid that refactoring to be honest.

akabiru commented on modules/storages/app/controllers/storages/admin/storages/project_storages_controller.rb:
```diff
@@ -73,10 +72,24 @@ def create
+  end
+
+  def destroy
+    @project_storage.destroy
```
🟠 There's an edge case where a user can attempt to unlink a project from different tabs- in which case they would be met by an unhandled 404- in project attributes, we handled this via a before action [here](https://github.com/opf/openproject/blob/dev/app/controllers/admin/settings/project_custom_fields_controller.rb#L186-L189) shall we extend the same?

----


https://github.com/user-attachments/assets/37fbee7f-5817-409e-a5cd-e391ad6184c4

akabiru replied:
See https://github.com/opf/openproject/pull/15653

judithroth replied:
Thanks for pointing that out!

akabiru commented on modules/storages/app/components/storages/project_storages/destroy_confirmation_dialog_component.html.erb:
```diff
@@ -0,0 +1,86 @@
+    url: admin_settings_storage_project_storage_path(id: @project_storage),
+    data: {
+      turbo: true,
+      controller: 'disable-when-checked',
```
Nice one, I didn't know we had this already 👍🏾

akabiru commented on modules/storages/spec/components/storages/project_storages/destroy_confirmation_dialog_component_spec.rb:
```diff
@@ -0,0 +1,75 @@
+      end
+    end
+  end
+end
```
✨

akabiru commented on modules/storages/spec/components/storages/project_storages/destroy_confirmation_dialog_component_spec.rb:
```diff
@@ -0,0 +1,75 @@
+  end
+
+  describe "#text" do
+    let(:storage) { create(:storage, :as_generic) }
```
?
```suggestion
    let(:storage) { build_stubbed(:storage, :as_generic) }
```

akabiru reviewed (changes requested):
Looking good! Just a few Qs and requests for changes

judithroth committed:
[#55968] Review feedback

- Ensure nice behaviour when a project is removed that was already
  removed (e.g. in a different browser tab)
- prefer `build_stubbed` over `create` to speed up tests

judithroth committed:
[#55968] Improve spec

Instead of only checking if the page contains the project folder types
anywhere, it should also be ensured that they are in the expected table row.

An alternative to this would have been

```
expect(page).to have_text("#{project.name} Automatically managed", normalize_ws: true)
expect(page).to have_text("#{archived_project.name} No specific folder", normalize_ws: true)
```

But to make it explicit that we want to check that the contents of a
table row match our expectations, I added a new "within" method for
selecting table rows.

judithroth committed:
[#55968] Make spec more robust against flickering
